[RAM] Window Maintenance Client/Saved Object and Mapping/REST APIs

JiaweiWu opened the pull request:
## Summary
Resolves: https://github.com/elastic/kibana/issues/152270
Specs: https://docs.google.com/document/u/1/d/1-QblF6P19W9o5-10Us3bfgN80GRfjSIhybHrvJS_ObA/edit

This PR implements the following:
- New maintenance window SO
- New maintenance window client in the alerting plugin (generates and queries maintenance window events, and other CRUD functionality around the SO)
- New maintenance window REST APIs 
- Kibana privileges for reading/writing maintenance window

This PR does not include integration with task runner, a new PR will be created to do that work. 
 
## APIs:

```
Find all maintenance windows in current space
POST `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window/_find`
body: {}
```

```
Create maintenance window:
POST `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window`
body: {
  title: string,
  duration: number,
  r_rule: RRule
}
```

```
Update maintenance window by ID:
POST `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window/{id}`,
body: {
  title?: string,
  duration?: number,
  enabled?: boolean,
  r_rule?: RRule,
}
```

```
Get maintenance window by ID:
GET `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window/{id}`,
```

```
Delete maintenance window by ID:
DELETE `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window/{id}`,
```

```
Archive maintenance window by ID:
POST `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window/{id}/_archive`,
body: {
  archive: boolean
}
```

```
Finish maintenance window by ID:
POST `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window/{id}/_finish`,
```


## Maintenance window response schema:
```
{
  id: string;
  title: string;
  enabled: boolean;
  duration: number;
  expirationDate: string;
  events: DateRange[];
  rRule: RRuleParams;
  status: 'running' | 'upcoming' | 'finished' | 'archived';
  startDate: string | null;
  endDate: string | null;
  createdBy: string | null;
  updatedBy: string | null;
  createdAt: string;
  updatedAt: string;
}
```

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

JiaweiWu commented:
@elasticmachine merge upstream

doakalexi reviewed (approved):
LGTM!

JiaweiWu commented on x-pack/plugins/alerting/common/maintenance_window.ts:
```diff
@@ -0,0 +1,60 @@
+
+export type MaintenanceWindow = MaintenanceWindowSOAttributes & {
+  status: MaintenanceWindowStatus;
+  startDate: string | null;
+  endDate: string | null;
```
We have these start and end dates that map to: 
![image](https://user-images.githubusercontent.com/74562234/228920165-befd8b77-141f-4c52-a1c7-7a708af4717c.png)

The dates indicate the start and end date of the current or upcoming EVENT, not the start and end date of the entire maintenance window schedule, should I name these something like `eventStartDate` instead?

doakalexi replied:
`eventStartDate` sounds good to me, I agree that start and end date could be confusing

JiaweiWu commented:
@elasticmachine merge upstream

jeramysoucy commented on x-pack/plugins/alerting/server/saved_objects/index.ts:
```diff
@@ -125,6 +129,13 @@ export function setupSavedObjects(
+  savedObjects.registerType({
+    name: MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE,
+    hidden: true,
+    namespaceType: 'single',
```
I'm not too familiar with alerting and related features and the scope of use cases, but just wanted to make sure we're ok with this being a single namespace type.

JiaweiWu replied:
Yes this is a good point, we don't foresee any reason to make the maintenance window SO accessible across spaces but I will chat with a few more folks on the team just to be sure. Thanks for pointing this out.

legrego replied:
@JiaweiWu Even if you think you won't use it, assigning a namespace type of `multiple-isolated` is probably worth doing. This will give you the space isolation of `single`, but it will also leave the possibility open to making this sharable at any point in the future, without having to wait for a painful Breaking Change event.

JiaweiWu replied:
Sounds good, I will use `multiple-isolated` in that case :smile:

jeramysoucy reviewed (approved):
Looks good! I just had one question regarding the SO namespace type.

JiaweiWu commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

jeramysoucy commented on x-pack/plugins/alerting/server/saved_objects/index.ts:
```diff
@@ -125,6 +129,13 @@ export function setupSavedObjects(
+  savedObjects.registerType({
+    name: MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE,
+    hidden: true,
+    namespaceType: 'multiple-isolated',
```
👍

jeramysoucy reviewed (approved):
Looks good from our perspective! Thanks for making the SO namespace type change!

JiaweiWu commented:
@elasticmachine merge upstream

JiaweiWu commented:
@elasticmachine merge upstream

XavierM commented on x-pack/plugins/alerting/server/maintenance_window_client/get_maintenance_window_date_and_status.ts:
```diff
@@ -0,0 +1,131 @@
+}
+
+// Returns the most recent/relevant event and the status for a maintenance window
+export const getMaintenanceWindowDateAndStatus = ({
```
it looks like this function is not used any where, is that correct?

JiaweiWu replied:
It's being used in `getMaintenanceWindowFromRaw`, basically we use this helper to decorate the maintenance window SO with the current `status`, the `event_start/end_dates` when returned by the API, kind of similar to `getAlertFromRaw`

XavierM commented on x-pack/plugins/alerting/server/saved_objects/maintenance_window_mapping.ts:
```diff
@@ -0,0 +1,52 @@
+import { SavedObjectsTypeMappingDefinition } from '@kbn/core/server';
+import { rRuleMappingsField } from './rrule_mappings_field';
+
+export const maintenanceWindowMappings: SavedObjectsTypeMappingDefinition = {
```
Let's go over together but I do not think that we need add this attributes to be indexed, only the one we are going to have sort/filter/search on it. Since we are going to do all this feature in memory of the UI, I do not think that we need anything. what do you think?

JiaweiWu replied:
Yea sounds good to me, I'm not sure how indexing works behind the scenes, but if we need to index it again, can we add it back to the mapping?

JiaweiWu replied:
I will remove the explicit mappings for all fields except for `events` and `enabled` since we filter on those.

XavierM commented on x-pack/plugins/alerting/server/routes/lib/rrule_schema.ts:
```diff
@@ -0,0 +1,48 @@
+import { createValidateRruleBy } from '../../lib/validate_rrule_by';
+import { validateSnoozeStartDate, validateSnoozeEndDate } from '../../lib/validate_snooze_date';
+
+export const rRuleSchema = schema.object({
```
That's great to have!

XavierM commented on x-pack/plugins/alerting/server/maintenance_window_client/methods/archive.ts:
```diff
@@ -0,0 +1,99 @@
+  context: MaintenanceWindowClientContext,
+  params: ArchiveParams
+): Promise<MaintenanceWindow> {
+  return await retryIfConflicts(
```
I am wondering if we really need to do that or just use the saved object options `retryOnConflict`. They already doing it for us what do you think?

JiaweiWu commented on x-pack/plugins/alerting/server/maintenance_window_client/methods/update.ts:
```diff
@@ -0,0 +1,110 @@
+    });
+
+    if (!shouldRegenerateEvents({ maintenanceWindow: attributes, rRule, duration })) {
+      events = mergeEvents({ oldEvents: attributes.events, newEvents: events });
```
The main reason is to make sure we do not overwrite events that have been finished early. When we finish an event, we modify that event entry's end date to be `now`, but if we edit the rule and just regenerate everything, it would overwrite that finished event and would go back to being `running`.

The more robust way of doing this is to keep another array of modified events (I think google calendar does this).

JiaweiWu commented:
In the last commit https://github.com/elastic/kibana/pull/153411/commits/d65b6a9ce60266b9b336fd5d321f99feb725564b, I made the following changes:

- Remove 'filter' parameter for the `find` endpoint, changed the HTTP route from a `POST` to `GET`
- Modified `mergeEvent` function to merge only modified events
- Changed maintenance window mapping by removing most of the properties that we do not sort/filter on
- Move maintenance window routes to its own folder
- Update any associated tests

XavierM reviewed (approved):
Code looks good and working as expected. Please create an issue tech debt to bring back all the maintenance window files under one folder `alerting/server/maintenance_window`

Awesome work here!!!

JiaweiWu commented:
@elasticmachine merge upstream
